xmview-os/um_cmd/um_tag.1
.\" Copyright (c) 2012 Renzo Davoli
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License,
.\" version 2, as published by the Free Software Foundation.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
.\" MA 02110-1301 USA.

.TH VIEW-OS 1 "May 30, 2009" "VIEW-OS: a process with a view"
.SH NAME
um_tag \- View-OS tagged mount management
.SH SYNOPSIS
.B um_tag set 
.I tagset
.I command
[
.I args
]
.br
.B um_tag add
.I tagset
.I command
[
.I args
]
.br
.B um_tag del
.I tagset
.I command
[
.I args
]
.br
.B um_tag get
.br
.B um_tag lget
.br
.B um_tag list
.I tagset
.br
.B um_tag llist
.I tagset
.br
.B um_tag listall
.br
.B um_tag setname
.I tag
.I name
.br
.B um_tag getname
.I tag
.br
.SH DESCRIPTION
This command permits the management of the ViewOS tagged mount feature.

Each process has its own set of tags. The current implementation of View-OS supports
32 tags, numbered 0 to 31. The default set of tags includes all the tags.
The set of task of a process is inherited through fork and exec, i.e. when a process
has a specific set of tags, all the processes it generates normally have the same
set of tags.
Each "mounted" file system or service is visible for the processes whose set ot tags
intersects the set of tags of the process who did the mount.
In other words, given the set of tags A used for the mount, that mount will be "visible" by
all the processes which share at least one tag with A.

This command permits to start a command having a different set of tags.
For example
.br
\fBum_tag set 1,2,3,4 bash\fR
.br
this commands starts a shell whose set of tags is {1,2,3,4}.

It is also possible to assign names to tags.
For example
.br
\fBum_tag setname 1 foo
.br
\fBum_tag setname 2 bar
.br
\fBum_tag set foo,bar bash
\fR
.br
this commands starts a shell whose set of tags is foo,bar {1,2}.

It is also possible to write the tagset as a hexadecimal number:
.br
\fBum_tag set 0x1e bash\fR
\fR
.br
this commands starts a shell whose set of tags is {1,2,3,4}.
(i-th tag corresponds to the i-th least significant bit:
0 is 0x1, 1 is 0x2, 2 is 0x4 and so on).

.SH OPTIONS
.TP
.B um_tag set \fI tagset command [ args ] \fR
.br
set the set of tags for the command.

.TP
.B um_tag add \fI tagset command [ args ] \fR
.br
set the set of tags for the command, add the tags of the tagset parameter.

.TP
.B um_tag del \fI tagset command [ args ] \fR
.br
set the set of tags for the command, delete the tags of the tagset parameter.

.TP
.B um_tag get
.br
show the current tagset (the output is a hexadecimal number)

.TP
.B um_tag lget
.br
show the current tagset (the output is a list of tags)

.TP
.B um_tag list \fI tagset \fR
.br
show the equivalent list of tags of tagset.

.TP
.B um_tag llist \fI tagset \fR
.br
show the equivalent list of tags of tagset, the output shows the mapping between tags and bits.

.TP
.B um_tag listall
.br
list all the tags and their name.

.TP
.B um_tag setname \fI tag name \fR
.br
assign a name to a tag.

.TP
.B um_tag getname \fI tag \fR
.br
return the tag's name.

.SH SEE ALSO
.BR umview(1), 
.BR kmview(1)
.SH AUTHORS
View-OS is a project of the Computer Science Department, University of
Bologna. Project Leader: Renzo Davoli. 
.br
<http://www.sourceforge.net/projects/view-os>

Howto's and further information can be found on the project wiki
<wiki.virtualsquare.org>.



0.1/NEWS
Bug fixed in fat_write_data
Bug fixed in fat_mkdir (zeroing of new allocated clusters)

xmview-os/um_cmd/um_tag.c
/*   This is part of um-ViewOS
 *   The user-mode implementation of OSVIEW -- A Process with a View
 *
 *   um_tag.c 
 *   management of mount tags.
 *   
 *   Copyright 2009 Renzo Davoli University of Bologna - Italy
 *   
 *   This program is free software; you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License, version 2, as
 *   published by the Free Software Foundation.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License along
 *   with this program; if not, write to the Free Software Foundation, Inc.,
 *   51 Franklin St, Fifth Floor, Boston, MA  02110-1301 USA.
 *
 *   $Id: viewname.c 464 2008-04-17 10:53:55Z garden $
 *
 */   
#include <config.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <getopt.h>
#include <config.h>
#include <errno.h>
#include <ctype.h>
#include <um_lib.h>

char *tagname[32];

void usage()
{
	fprintf(stderr, 
			"Usage:\n"
			"\tum_tag set tagset command args\n"
			"\tum_tag add tagset command args\n"
			"\tum_tag del tagset command args\n"
			"\tum_tag get\n"
			"\tum_tag lget\n"
			"\tum_tag list tagset\n"
			"\tum_tag llist tagset\n"
			"\tum_tag listall\n"
			"\tum_tag setname tag name\n"
			"\tum_tag getname tag\n"
			"\n"
			"Management of View-OS mount tags\n"
			"\n");
	exit(1);
}

char *command_tags[]={"set","add","del","get","lget","list","llist","listall","setname","getname",NULL};
char  command_argc[]={    4,    4,    4,    2,     2,     3,     3,        2,        4,        3,   0};
enum command         { set,  add,  del,  get,  lget,  list,  llist,  listall,  setname,  getname, error};

int str2tag_single(char *s) {
	int tag;
	if (isdigit(*s))
		tag=atoi(s);
	else {
		for (tag=0; tag<32; tag++)
			if (strcmp(s,tagname[tag]) == 0)
				break;
	}
	if (tag < 0 || tag >= 32)
		fprintf(stderr,"Tag '%s' unknown\n",s);
	return tag;
}

int str2tags_single (char *s){
	int tags=0;
	if (strncmp(s,"0x",2) == 0) {
		sscanf(s+2,"%x",&tags);
	} else
		tags = 1<<str2tag_single(s);
	//printf("str2tags_single  %s %x\n",s,tags);
	return tags;
}

int str2tags (char *s) {
	char *next;
	int tags=0;
	while((next=strtok(s,",")) != NULL) {
		tags |= str2tags_single(next);
		s=NULL;
	}
	return tags;
}

int str2tag (char *s)
{
	if (strncmp(s,"0x",2) == 0) {
		int tags,tag;
		sscanf(s+2,"%x",&tags);
		for (tag=0; tag<32; tag++) {
			if (tags & (1<<tag)) {
				tags &= ~(1<<tag);
				if (tags) 
					fprintf(stderr,"'%s' does not refer to a single tag. Using '0x%x' instead\n",1<<tag);
				return tag;
			}
		}
		return 0;
	} else
		return str2tag_single(s);
}

enum command getcommand(char *command)
{
	int i;
	while (command_tags[i] != NULL && strcmp(command_tags[i],command) != 0)
		i++;
	return (enum command) i;
}

main(int argc, char *argv[])
{
	int tag;
	enum command cmd;
	int tags; 
	int rv=0;
	if (argc < 2)
		usage();
	for (tag=0; tag<32; tag++) {
		char buf[256];
		rv=um_tagstring(VIEWOS_TAGSTRING_GET,tag,buf,256);
		tagname[tag]=strdup(buf);
	}
	cmd=getcommand(argv[1]);
	if (argc < command_argc[cmd])
		usage();
	if (cmd>2 && argc > command_argc[cmd])
		usage();
	switch (cmd) {
		case set:
			tags=str2tags(argv[2]);
			//printf("%08x\n",tags);
			rv=um_tag(VIEWOS_TAG_SET, &tags, NULL, sizeof(int));
			execvp(argv[3],argv+3);
			break;
		case add:
			tags=str2tags(argv[2]);
			//printf("%08x\n",tags);
			rv=um_tag(VIEWOS_TAG_ADD, &tags, NULL, sizeof(int));
			execvp(argv[3],argv+3);
			break;
		case del:
			tags=str2tags(argv[2]);
			//printf("%08x\n",tags);
			rv=um_tag(VIEWOS_TAG_DEL, &tags, NULL, sizeof(int));
			execvp(argv[3],argv+3);
			break;
		case get:
			rv=um_tag(VIEWOS_TAG_SET, NULL, &tags, sizeof(int));
			if (rv == 0)
				printf("0x%x\n",tags);
			break;
		case lget: 
			rv=um_tag(VIEWOS_TAG_SET, NULL, &tags, sizeof(int));
			if (rv == 0) {
				char *sep="";
				for (tag=0; tag<32; tag++) {
					if (tags & 1<<tag) {
						if (tagname[tag][0])
							printf("%s%s",sep,tagname[tag]);
						else
							printf("%s%02d",sep,tag);
						sep=",";
					}
				}
				printf("\n");
			}
			break;
		case list: 
			tags=str2tags(argv[2]);
			char *sep="";
			for (tag=0; tag<32; tag++) {
				if (tags & 1<<tag) {
					if (tagname[tag][0])
						printf("%s%s",sep,tagname[tag]);
					else
						printf("%s%02d",sep,tag);
					sep=",";
				}
			}
			printf("\n");
			break;
		case llist: 
			tags=str2tags(argv[2]);
			printf("   0x%08x\n",tags);
			for (tag=0; tag<32; tag++) {
				if (tags & 1<<tag) {
					printf("%02d 0x%08x %s\n",tag,1<<tag,tagname[tag]);
				}
			}
			break;
		case listall:
			for (tag=0; tag<32; tag++) {
				if (tagname[tag][0]) {
					printf("%02d 0x%08x %s\n",tag,1<<tag,tagname[tag]);
				}
			}
			break;
		case setname:
			tag=str2tag(argv[2]);
			um_tagstring(VIEWOS_TAGSTRING_SET,tag,argv[3],strlen(argv[3]));
			break;
		case getname:
			tag=str2tag(argv[2]);
			printf("%02d 0x%08x %s\n",tag,1<<tag,tagname[tag]);
			break;
	}
	if (rv < 0) {
		fprintf(stderr, "Error: %s\n", strerror(errno));
		exit(1);
	}
	exit (0);
}
